### PR TITLE
replaced spaghetti js with react

### PR DESCRIPTION
**Pull request by ishmandoo**

### CODE REVIEW

File: server/static/hardware_list.jsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+const e = React.createElement;
+
+class HardwareList extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = { hardware: [] };
+    }
+
+    componentDidMount() {
```
Comment (MaximilianKing) [inline server/static/hardware_list.jsx:11]: What does this do? Is this the new heartbeat stuff?

File: server/static/hardware_list.jsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        this.update()
+        this.timerID = setInterval(
+            () => this.update(),
+            1000
+        );
+    }
+
+    update() {
```
Comment (MaximilianKing) [inline server/static/hardware_list.jsx:19]: Is this setting the hardware state? if so, would it be better to have this just get the hardware state and we can maintain it in the flask app? just curious, would love to know how this works more.

File: server/static/hardware_list.jsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                    <span className={`float-right badge ${badge_type}`}>{hardware.status}</span>
+                </td>
+            </tr >
+        )
+
+    }
+
+    render() {
+        const listItems = this.state.hardware.map((hardware) => this.get_hardware_row(hardware));
+        console.log(listItems)
```
Comment (MaximilianKing) [inline server/static/hardware_list.jsx:63]: Is this console.log going to log it to std out or is this react way of printing to a template/page?

File: server/static/job_list.jsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+const e = React.createElement;
+
+class JobList extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = { queued: [], running: [], completed: [] };
+    }
+
+    componentDidMount() {
```
Comment (MaximilianKing) [inline server/static/job_list.jsx:11]: This looks very familiar with the above in hardware. Is it possible to make abstract base classes for these React elements and classes?

File: server/static/job_list.jsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+                </td>
+            </tr >
+        )
+
+    }
+
+    render() {
+        var jobs = this.state.queued.concat(this.state.running, this.state.completed)
+        const listItems = jobs.map((job) => this.get_job_row(job));
+        console.log(listItems)
```
Comment (MaximilianKing) [inline server/static/job_list.jsx:86]: Same as the above console.log question.

File: server/static/submit.jsx [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        this.handleChange = this.handleChange.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+    }
+
+    resetState(error) {
+        this.setState({ git_url: '', git_user: '', project_name: '', valid: false, error: error });
+    }
+
+    handleChange(event) {
+        var re = RegExp('^(?:https:\/\/|http:\/\/|)(?:www.|)github.com\/([a-z0-9]*)\/([a-z0-9-_]*)(?:\/|)(?:.git|)$')
```
Comment (MaximilianKing) [inline server/static/submit.jsx:20]: Yikes. I am guessing we need to use regex here?

File: server/static/submit.jsx [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+                            error: "repo_not_found"
+                        });
+                    }
+                }
+            ).catch(
+                (error) => {
+                    this.setState({
+                        valid: false,
+                        error: "unknown"
+                    });
```
Comment (MaximilianKing) [inline server/static/submit.jsx:74]: We kill this?

File: server/static/submit.jsx [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+        fetch('/job', {
+            method: 'POST',
+            body: form_data,
+        });
+        window.document.location = "/"
+    }
+
+    render() {
+        return (
```
Comment (MaximilianKing) [inline server/static/submit.jsx:108]: All this render return jazz, can it be held somewhere else like we did with the html templates? Or is that dumb? Just thinking that this is very formatty of our content type code where the rest is queries and backend shit. Just to try and divide out backend vs frontend. Just a thought, not a real strong suggestion.

File: server/templates/header.html [hunk 9]
```diff
@@ -27,4 +29,15 @@
 <div class="container">
         <a href="/"><img src="/static/banner_small.png" class="img-fluid"><img></a>
 </div>
+
+
+
+<!-- Load React. -->
+<!-- Note: when deploying, replace "development.js" with "production.min.js". -->
```
Comment (MaximilianKing) [inline server/templates/header.html:36]: What does this comment mean? We need to replace a file when we want to `eb deploy`?

File: server/application.py [hunk 10]
```diff
@@ -255,6 +256,7 @@ def submit_page_route():
 def job_route():
     jobs_cache.get_db_cache()
     if request.method == "POST":
+        print(request)
```
Comment (MaximilianKing) [inline server/application.py:259]: We want this for cloudwatch logs or is this a temp debugging thing?

File: server/static/hardware_list.jsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
+'use strict';
```
Comment (MaximilianKing) [inline server/static/hardware_list.jsx:1]: Splain me! I wants to learn.

File: server/static/hardware_list.jsx (hunk 1 above)
Comment (ishmandoo) [reply server/static/hardware_list.jsx:11]: This just polls the server for changes in the hardware status.

File: server/application.py (hunk 10 above)
Comment (ishmandoo) [reply server/application.py:259]: Oops. Temporary debugging. I'll remove it.

File: server/static/hardware_list.jsx (hunk 11 above)
Comment (ishmandoo) [reply server/static/hardware_list.jsx:1]: Tells the JS intepreter to use strict mode. The only difference I know of (though I'm sure there are others) is that it doesn't let you introduce new variables without the `var` keyword. In non-strict javascript, if you declare a variable with just
```
x = 100
```
you get a _global_ variable for some reason. The programmer almost always means
```
var x = 100
```
which makes a local variable. Strict mode forces you to declare variables with `var`.

File: server/static/hardware_list.jsx (hunk 2 above)
Comment (ishmandoo) [reply server/static/hardware_list.jsx:19]: This sets the state of the React component. For this component, that state is just the last known status of all hardware.

File: server/static/hardware_list.jsx (hunk 3 above)
Comment (ishmandoo) [reply server/static/hardware_list.jsx:63]: Going to the browser's console. I should remove this.

File: server/static/job_list.jsx (hunk 4 above)
Comment (ishmandoo) [reply server/static/job_list.jsx:11]: I'm sure there's a way to share code like this. I don't know it, though.

File: server/static/job_list.jsx (hunk 5 above)
Comment (ishmandoo) [reply server/static/job_list.jsx:86]: This should go too.

File: server/static/submit.jsx (hunk 6 above)
Comment (ishmandoo) [reply server/static/submit.jsx:20]: This regex parses github repo urls. If there's another way to do it, I'm open to it.

File: server/static/submit.jsx (hunk 7 above)
Comment (ishmandoo) [reply server/static/submit.jsx:74]: Not sure what to do here. Checking of the `run.py` exists in the repo is a nice feature, but while testing I ran into the API's query limit. I suspect some users will too.

Currently, it hits the GitHub API after every character. We could make it only checks after the user has stopped typing for a few seconds. We could show the user a spinner during those seconds to show that it's thinking. 

Either way we should probably have the React app catch the Github API query limit error and treat it as if `run.py` exists. That way the user gets this protection at least some of the time. 

Or we could just scrap this feature.

File: server/static/submit.jsx (hunk 8 above)
Comment (ishmandoo) [reply server/static/submit.jsx:108]: Mixing JS and HTML seems to be the React way. I think?

File: server/templates/header.html (hunk 9 above)
Comment (ishmandoo) [reply server/templates/header.html:36]: We should talk about this. Not quite sure.

**Pull request closed without merging**